### PR TITLE
Move most dependencies to `extra_requires` except bare-minimum deps

### PR DESCRIPTION
**Pull request by zpapakipos**
## Description
I modified the `setup.py` file by moving most of the dependencies in `install_requires` to separate `extra_requires`. This will lighten the installation process for most users as they can only install the dependencies needed for the parts of the `vidgear` API that they will use. Some example usages will look like the following:
```
pip install vidgear  # installs bare-minimum deps
```
```
pip install vidgear[camgear]  # installs bare-minimum + deps needed for camgear API
```
```
pip install vidgear[netgear, asyncio]  # installs bare-minimum + deps needed for netgear & asyncio APIs
```

@abhiTronix can you please checkout this branch, test installing & building the wheel using this version of the `setup.py`, and once you verify it's working upload a new version of `vidgear` to pypi so that we can install using these bare-minimum deps? Thanks! :)

### Requirements / Checklist

- [x] I have read the [PR Guidelines](https://abhitronix.github.io/vidgear/latest/contribution/PR/#submitting-pull-requestpr-guidelines).
- [x] I have read the [Documentation](https://abhitronix.github.io/vidgear/latest).
- [x] I have updated the documentation accordingly(if required).

### Related Issue

https://github.com/abhiTronix/vidgear/issues/267

### Context

As I said above, this will lighten the installation process for most users as they can only install the dependencies needed for the parts of the `vidgear` API that they will use. For us for example this will unblock us so we are able to install the bare-minimum `vidgear` in our codebase which is already quite heavy (so installing unnecessary deps is a pain).

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by abhiTronix** [comment]
@zpapakipos Sorry but you have to change base branch from `master` to `testing` branch(as mentioned in our [contribution guidelines](https://abhitronix.github.io/vidgear/latest/contribution/PR/#clone-testing-branch)), as we don't allow changes directly into `master` branch. Then only we can work with it.

See this doc https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request for doing it easily,

**Comment by zpapakipos** [comment]
@abhiTronix done!

**Comment by zpapakipos** [comment]
@abhiTronix I resolved the merge conflicts -- it looks like in the testing branch the `colorlog` dependency had been modified so it required the most recent version. I don't think this should be necessary, so I removed that change. Let me know if there is some reason `colorlog` needs to be **the most recent** version for `vidgear` to run. This would complicate things a lot on our side.

**Comment by abhiTronix** [comment]
> @abhiTronix I resolved the merge conflicts

@zpapakipos Thank you again.

>   it looks like in the testing branch the colorlog dependency had been modified so it required the most recent version. I don't think this should be necessary, so I removed that change. Let me know if there is some reason colorlog needs to be the most recent version for vidgear to run.

No, The recent version is added to colorlog just to include more notable changes in its latest release which provide better formatting.

> This will lighten the installation process for most users as they can only install the dependencies needed for the parts of the vidgear API that they will use. Some example usages will look like the following:
pip install vidgear  # installs bare-minimum deps
pip install vidgear[camgear]  # installs bare-minimum + deps needed for camgear API
pip install vidgear[netgear, asyncio]  # installs bare-minimum + deps needed for netgear & asyncio APIs

I think this will complicate installation for everyday user. In my opinion,  the correct way to do this is to bundling all API specific dependencies under one name such as "core" and also providing same dependencies under "asyncio" as well(since `extra_requires` doesn't share dependencies in different names) along with its specific dependencies. So final configrations would be:

```sh
pip install vidgear  # installs bare-minimum deps
pip install vidgear[core]  # installs bare-minimum deps + API specific deps
pip install vidgear[asyncio]  # installs bare-minimum deps + API specific deps + Asyncio deps
```

**Comment by abhiTronix** [comment]
Also we don't need recent version for all deps, it's for selected ones that need updation more frequently for its working properly like `yt_dlp`. But leave it, as I'll correct it later.

File: setup.py [hunk 1]
```diff
@@ -1,186 +1,192 @@
...
+        "camgear": [
+            "pafy{}".format(latest_version("pafy")),
+            "yt_dlp{}".format(latest_version("yt_dlp")),  # pafy backend
+            "streamlink",
+        ],
+        "netgear": [
+            "pyzmq{}".format(latest_version("pyzmq")),
+            "simplejpeg{}".format(latest_version("simplejpeg")),
+        ],
+        "screengear": [
```
Comment (zpapakipos) [inline setup.py:136]:
```suggestion
        "core": [
            "pafy{}".format(latest_version("pafy")),
            "yt_dlp{}".format(latest_version("yt_dlp")),  # pafy backend
            "streamlink",
            "pyzmq{}".format(latest_version("pyzmq")),
            "simplejpeg{}".format(latest_version("simplejpeg")),
```

File: setup.py [hunk 2]
```diff
@@ -1,186 +1,192 @@
...
+    ]
+    + (["opencv-python"] if test_opencv() else [])
+    + (["picamera"] if ("arm" in platform.uname()[4][:3]) else []),
+    long_description=long_description,
+    long_description_content_type="text/markdown",
+    author_email="abhi.una12@gmail.com",
+    url="https://abhitronix.github.io/vidgear",
+    extras_require={
+        "asyncio": [
+            "starlette{}".format(latest_version("starlette")),
```
Comment (zpapakipos) [inline setup.py:111]:
```suggestion
            "pafy{}".format(latest_version("pafy")),
            "yt_dlp{}".format(latest_version("yt_dlp")),  # pafy backend
            "streamlink",
            "pyzmq{}".format(latest_version("pyzmq")),
            "simplejpeg{}".format(latest_version("simplejpeg")),
            "mss{}".format(latest_version("mss")),
            "Pillow",
            "pyscreenshot{}".format(latest_version("pyscreenshot")),
            "starlette{}".format(latest_version("starlette")),
```

**Comment by zpapakipos** [comment]
> I think this will complicate installation for everyday user. In my opinion, the correct way to do this is to bundling all API specific dependencies under one name such as "core" and also providing same dependencies under "asyncio" as well(since extra_requires doesn't share dependencies under different names) along with its own dependencies.

This sounds good to me! I made the change to this branch, please take a look & let me know if it looks good :)

> No, The recent version is added to colorlog just to include more notable changes in its latest release which provide better formatting.

I see. Is it ok with you if we put just `colorlog` in the base requirements and then if people want to have the most up-to-date log formatting they can choose to update `colorlog` themselves? Or alternatively you could put `colorlog` with the most recent version in the `core` requirement extra, but only `colorlog` in the base requirements. For example for us, we don't want to have to install a whole new version of `colorlog` just so we can install `vidgear` in our internal codebase (we already have 3.1.4 installed, which is old but works fine for our use case).

**Comment by abhiTronix** [comment]
> I see. Is it ok with you if we put just colorlog in the base requirements and then if people want to have the most up-to-date log formatting they can choose to update colorlog themselves? Or alternatively you could put colorlog with the most recent version in the core requirement extra, but only colorlog in the base requirements. For example for us, we don't want to have to install a whole new version of colorlog just so we can install vidgear in our internal codebase (we already have 3.1.4 installed, which is old but works fine for our use case).

@zpapakipos Sounds good to me. 👍🏽 

I've suggested few changes above, make sure to commit them as well, and also don't worry above test failing as they're caused by old `pafy` dependency that's subjected to replaced/removed soon.

**Comment by zpapakipos** [comment]
I don't see any changes you suggested. Can you comment on the code with any changes? Then I'm happy to add them.

Also, once this is landed how soon do you plan to release v0.2.4? For our use it would be great to have the new version in the next week, is that possible?

**Comment by abhiTronix** [comment]
> I don't see any changes you suggested. Can you comment on the code with any changes? Then I'm happy to add them.

@zpapakipos See reviews left on the code above:
image

**Comment by abhiTronix** [comment]
> Also, once this is landed how soon do you plan to release v0.2.4? For our use it would be great to have the new version in the next week, is that possible?

Very much possible, but I first need to update docs and replace `pafy` dependency.

**Review by abhiTronix — CHANGES_REQUESTED** [review]
@zpapakipos Please make these changes.

File: setup.py [hunk 3]
```diff
@@ -1,186 +1,196 @@
...
+    author="Abhishek Thakur",
+    install_requires=[
+        "cython",  # helper for numpy install
+        "numpy",
+        "requests",
+        "colorlog",
+        "tqdm",
+    ]
+    + (["opencv-python"] if test_opencv() else [])
+    + (["picamera"] if ("arm" in platform.uname()[4][:3]) else []),
```
[the remark this answers is not in the document]
Comment (zpapakipos) [reply setup.py:104]:
```suggestion
```

File: setup.py [hunk 4]
```diff
@@ -1,186 +1,196 @@
...
+            "mss{}".format(latest_version("mss")),
+            "Pillow",
+            "pyscreenshot{}".format(latest_version("pyscreenshot")),
+            "starlette{}".format(latest_version("starlette")),
+            "jinja2",
+            "uvicorn{}".format(latest_version("uvicorn")),
+            "msgpack{}".format(latest_version("msgpack")),
+            "msgpack_numpy{}".format(latest_version("msgpack_numpy")),
+            "aiortc{}".format(latest_version("aiortc")),
+        ]
```
Comment (zpapakipos) [inline setup.py:125]:
```suggestion
        ]
        + (["picamera"] if ("arm" in platform.uname()[4][:3]) else []),
```

File: setup.py [hunk 5]
```diff
@@ -1,186 +1,196 @@
...
+        "core": [
+            "pafy{}".format(latest_version("pafy")),
+            "yt_dlp{}".format(latest_version("yt_dlp")),  # pafy backend
+            "streamlink",
+            "pyzmq{}".format(latest_version("pyzmq")),
+            "simplejpeg{}".format(latest_version("simplejpeg")),
+            "mss{}".format(latest_version("mss")),
+            "Pillow",
+            "pyscreenshot{}".format(latest_version("pyscreenshot")),
+        ],
```
Comment (zpapakipos) [inline setup.py:144]:
```suggestion
        ]
        + (["picamera"] if ("arm" in platform.uname()[4][:3]) else []),
```

File: setup.py [hunk 6]
```diff
@@ -1,186 +1,196 @@
...
+        ]
+        + (
+            (
+                ["uvloop{}".format(latest_version("uvloop"))]
+                if sys.version_info[:2] >= (3, 7)  # dropped support for 3.6.x legacies
+                else ["uvloop==0.14.0"]
+            )
+            if (platform.system() != "Windows")
+            else []
+        ),
```
[the remark this answers is not in the document]
Comment (zpapakipos) [reply setup.py:134]: Sorry, I'm not sure exactly where you want this -- can you add it as a suggestion & commit it to add to this branch yourself? I added the rest so let me know if they look good!

**Comment by zpapakipos** [comment]
> Very much possible, but I first need to update docs and replace `pafy` dependency.

Super glad to hear this, thanks! :)

I made/replied to all your changes, let me know if it looks good!

File: setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhiTronix) [reply setup.py:134]: @zpapakipos Sure.

**Review by abhiTronix — APPROVED** [review]
LGTM!

**Comment by zpapakipos** [comment]
Awesome, thanks for the quick turnaround time! :) Looking forward to vidgear v0.2.4 with these changes!

**Comment by abhiTronix** [comment]
@zpapakipos  **The latest vidgear `v0.2.4` is now available on pypi with this fix: https://pypi.org/project/vidgear/**. Goodluck 👍🏽

**Pull request merged**